### PR TITLE
cryptenroll: Use CTAP2.1 credProtect extension

### PR DESCRIPTION
**Pull request by BryanJacobs**
When enrolling a new FIDO2 token with a client PIN, this tells the authenticator to require the PIN on all uses.

Works around #31443 in most (not all) scenarios.

A full fix would require checking the `uv` flag of the result from the `getAssertion` call and comparing with the JSON token info for whether a PIN or UV was requested.

Nonetheless, this should be an improvement.

### CODE REVIEW

File: src/shared/libfido2-util.c [hunk 1]
```diff
@@ -776,10 +777,19 @@ int fido2_generate_hmac_hash(
...
+        int extensions = FIDO_EXT_HMAC_SECRET;
+        if (FLAGS_SET(lock_with, FIDO2ENROLL_PIN)) {
+                extensions |= FIDO_EXT_CRED_PROTECT;
+
+                r = sym_fido_cred_set_prot(c, FIDO_CRED_PROT_UV_REQUIRED);
+                if (r != FIDO_OK)
+                        log_notice("Failed to set protection level on FIDO2 credential: %s", sym_fido_strerr(r));
+        }
+
+        r = sym_fido_cred_set_extensions(c, extensions);
```
Comment (poettering) [inline src/shared/libfido2-util.c:792]: hmm, isn't this going to fail if FIDO_EXT_CRED_PROTECT is not supported?

Comment (BryanJacobs) [reply src/shared/libfido2-util.c:792]: I'll take a look at what libfido2 does, but on the authenticator (CTAP) side, what should happen is that the extension is ignored and the credential created without error. The result from the makeCredential wouldn't have credProtect in it.

All extensions are this way in CTAP - ignored by authenticators that don't support them.

Comment (BryanJacobs) [reply src/shared/libfido2-util.c:792]: https://github.com/Yubico/libfido2/blob/b63570ab677354cddee60075f522331fb1a66b0a/src/cred.c#L946 - no failure on the set_extensions call.

https://github.com/Yubico/libfido2/blob/b63570ab677354cddee60075f522331fb1a66b0a/src/cred.c#L91 - unconditionally shoved into the makeCredential payload by the library. ( https://github.com/Yubico/libfido2/blob/b63570ab677354cddee60075f522331fb1a66b0a/src/cbor.c#L627 no error on the encoding).

Read of the code says it should be okay. I'll test it end to end later today.

Comment (poettering) [reply src/shared/libfido2-util.c:792]: can you summarize this in a short comment in the code here. i.e. just mention that tokens are supposed to gracefully ignore our request to use the extension if they dont support it.

Comment (poettering) [reply src/shared/libfido2-util.c:792]: btw, a comment explaining briefly what use of the extension is supposed to achieve would be good anyway. i.e. some of the commit msg text should probably be a comment.

File: src/shared/libfido2-util.c [hunk 2]
```diff
@@ -40,6 +40,7 @@ DLSYM_FUNCTION(fido_cred_id_ptr);
 DLSYM_FUNCTION(fido_cred_new);
 DLSYM_FUNCTION(fido_cred_set_clientdata_hash);
 DLSYM_FUNCTION(fido_cred_set_extensions);
+DLSYM_FUNCTION(fido_cred_set_prot);
```
Comment (poettering) [inline src/shared/libfido2-util.c:43]: did you actually test this?

just definining the symbol should not be enough, you also need a prototype for it in the libfido2-util.h header, and then actually resolve it via a line in dlopen_libfido2()

Comment (BryanJacobs) [reply src/shared/libfido2-util.c:43]: I'm sorry, I assumed there would have been a unit test that broke if the dlopen failed. My bad :-(.

I've updated the PR and added a block that forces the presentation of a PIN prompt prior to credential creation when the user asks for a PIN-required credential. This fixes the underlying bug, #31443 .

I've also tested that it works locally:

```
✗ sudo -E ./buildDir/systemd-cryptenroll --fido2-device=auto --fido2-with-user-presence=false --fido2-with-client-pin=true /tmp/foo        
🔐 Please enter current passphrase for disk /tmp/foo: ••••••                  
Initializing FIDO2 credential on security token.
🔐 Please enter security token PIN: ••••••••••••            
Generating secret key on FIDO2 security token.
New FIDO2 token enrolled as key slot 3.
```

Without this change, there was no PIN prompt. With the first half of this change (adding credProtect=3), the get_assertion gets a response of NO_CREDENTIALS (as it should when systemd fails to collect the PIN!). With both halves of this change, the right hmac secret is used for the enrollment and all is well.

I don't have an authenticator to check the behavior when credProtect is unsupported - if you'd prefer if we do feature detection we could, but the standard is clear that authenticators should ignore unknown extensions: https://fidoalliance.org/specs/fido-v2.2-rd-20230321/fido-client-to-authenticator-protocol-v2.2-rd-20230321.html#op-makecred-step-extensions

> Process any extensions that this authenticator supports, ignoring any that it does not support.

File: src/shared/libfido2-util.c (hunk 1 above)
Comment (BryanJacobs) [reply src/shared/libfido2-util.c:792]: I've added comments describing the logic. Please let me know if they're not sufficiently clear.

File: src/shared/libfido2-util.c (hunk 2 above)
Comment (poettering) [reply src/shared/libfido2-util.c:43]: > I'm sorry, I assumed there would have been a unit test that broke if the dlopen failed. My bad :-(.

We actually do. But it wasn't triggered here, after all you never loaded the symbol, so it could not fail to load either...

File: src/shared/libfido2-util.c [hunk 3]
```diff
@@ -776,10 +778,21 @@ int fido2_generate_hmac_hash(
...
-        r = sym_fido_cred_set_extensions(c, FIDO_EXT_HMAC_SECRET);
+        int extensions = FIDO_EXT_HMAC_SECRET;
+        if (FLAGS_SET(lock_with, FIDO2ENROLL_PIN) || FLAGS_SET(lock_with, FIDO2ENROLL_UV)) {
+                /* Attempt to use the "cred protect" extension, requiring user verification (UV) for this
+                 * credential. If the authenticator doesn't support the extension, it will be ignored. */
+                extensions |= FIDO_EXT_CRED_PROTECT;
+
+                r = sym_fido_cred_set_prot(c, FIDO_CRED_PROT_UV_REQUIRED);
+                if (r != FIDO_OK)
+                        log_notice("Failed to set protection level on FIDO2 credential: %s", sym_fido_strerr(r));
```
Comment (poettering) [inline src/shared/libfido2-util.c]: if something fails but we ignore the failure and proceed we usually log at LOG_WARNING level, and we suffix the message with ", ignoring".

Hence, can you change this to:

```c
log_warning("Failed to set … credential, ignoring: %s", sym_fido_strerr(r));
```

**Comment by poettering** [comment]
looks good, just one minor nitpick, see above.

**Comment by poettering** [comment]
please squash.

ant log_warning, not log_notice, please

**Comment by BryanJacobs** [comment]
> please squash.
> 
> ant log_warning, not log_notice, please

Done. Apologies for wasting your time.

File: src/shared/libfido2-util.c [hunk 4]
```diff
@@ -830,7 +843,17 @@ int fido2_generate_hmac_hash(
...
+        /* If we are using the user PIN, then we must pass that PIN to the get_assertion call below, or
+         * the authenticator will use the non-user-verification HMAC secret (which differs from the one when
+         * the PIN is passed).
+         *
+         * Rather than potentially trying and failing to create the credential, just collect the PIN first
+         * and then pass it to both the make_credential and the get_assertion operations. */
+        if (FLAGS_SET(lock_with, FIDO2ENROLL_PIN))
+            r = FIDO_ERR_PIN_REQUIRED;
+        else
+            r = sym_fido_dev_make_cred(d, c, NULL);
```
Comment (poettering) [inline src/shared/libfido2-util.c]: 8 space indentation please, i.e follow coding style

Comment (poettering) [reply src/shared/libfido2-util.c]: still not fixed

Comment (BryanJacobs) [reply src/shared/libfido2-util.c]: I hadn't pushed the extra spaces on these two lines yet. It's fixed now.

**Comment by poettering** [comment]
thanks! lgtm!
